### PR TITLE
Implement the use of MNAIO images for system tests [pike]

### PR DESCRIPTION
**Pull request by odyssey4me**
This PR does the following:

1. To make testing the same way as the gate simpler outside of the gate, we implement a common vars script which is sourced by the pre/run/post stages, ensuring that all vars required for each stage are available.

2. Currently the tox tests are executed via multiple layers of scripts. This patch flattens it into the gating/check/run script to make following the code simpler.

3. We move the pre-installation of distro and python packages to the 'pre' stage to improve the chances of success, given that the 'pre' script is run three times before failing.

4. Currently the AIO deploy tests are executed via multiple layers of scripts. This patch flattens it into the gating/check/run script to make following the code simpler.

5. We move the MNAIO tempest run into run_deploy_mnaio.sh to reduce complexity and make the code easier to follow.

6. We cut down on duplicated var output between the 'prep' and 'run' stages.

7. Does all the MNAIO host preparation in the 'pre' stage of the gating pipeline. This stage is tried three times before failing, improving the chances of success. The actions done in this stage all relate to the VM preparation, and have nothing to do with deploying RPC-O.

8. Implements a check to verify whether there is an MNAIO image manifest available for this configuration. If so, it sets the preparation stage to download the images and use those to start the VM's.

9. If no images are available, or if this is a 'deploy' test, then build the VM's as usual.

10. Switches the 'system' tests to use the images if they are available.

Issue: [RE-1998](https://rpc-openstack.atlassian.net/browse/RE-1998)

Issue: [RE-1562](https://rpc-openstack.atlassian.net/browse/RE-1562)

### CODE REVIEW

File: gating/check/pre [hunk 1]
```diff
@@ -24,3 +26,27 @@ fi
...
+  dpkg-query --list | grep curl &>/dev/null || distro_pkgs_to_install+="curl "
+fi
+
+if [[ $RE_JOB_ACTION == "tox-test" ]]; then
+  dpkg-query --list | grep wget &>/dev/null || distro_pkgs_to_install+="wget "
+  pip freeze | grep tox &>/dev/null || py_pkgs_to_install+="tox==3.3.0 "
+fi
+
+if [ "${distro_pkgs_to_install}" != "" ]; then
+  apt-get install -y ${distro_pkgs_to_install}
```
Comment (hughsaunders) [inline gating/check/pre]: has an update definitely happened before this?

File: gating/check/run_deploy_mnaio.sh [hunk 2]
```diff
@@ -16,64 +16,47 @@
...
-# ssh command used to execute tests on infra1
-export MNAIO_SSH="ssh -ttt -oStrictHostKeyChecking=no root@infra1"
-# place variable in file to be sourced by parent calling script 'run'
-export MNAIO_VAR_FILE="${MNAIO_VAR_FILE:-/tmp/mnaio_vars}"
-echo "export MNAIO_SSH=\"${MNAIO_SSH}\"" > "${MNAIO_VAR_FILE}"
-echo "export RPC_RELEASE=\"${RPC_RELEASE}\"" >> "${MNAIO_VAR_FILE}"
-echo "export RPC_PRODUCT_RELEASE=\"${RPC_PRODUCT_RELEASE}\"" >> "${MNAIO_VAR_FILE}"
+source "$(readlink -f $(dirname ${0}))/../mnaio_vars.sh"
+
+# If there are images available, and this is not the 'deploy' action,
```
Comment (hughsaunders) [inline gating/check/run_deploy_mnaio.sh]: is or is not?

File: gating/check/run_system_tests.sh [hunk 3]
```diff
@@ -41,18 +40,17 @@ set +e
...
+fi
 
-# 4. Collect logs from script
+# 4. Collect logs from script, if they exist
 mkdir -p "${RE_HOOK_ARTIFACT_DIR}" || true  # ensure that artifact directory exists
-cp test_results.tar "${RE_HOOK_ARTIFACT_DIR}/molecule_test_results.tar"
-# Molecule does not produce logs outside of STDOUT
-# record non-zero exit code if not already recorded
-[[ $? -ne 0 ]] && [[ ! -z ${RC+x} ]] && RC=$?
+if [[ -e test_results.tar ]]; then
```
Comment (hughsaunders) [inline gating/check/run_system_tests.sh]: Comment says logs but this looks like results?

File: gating/check/pre (hunk 1 above)
Comment (odyssey4me) [reply gating/check/pre]: Yes, it happens in the ``gating_prerequisites.sh`` script on line 27.

File: gating/check/run_deploy_mnaio.sh (hunk 2 above)
Comment (odyssey4me) [reply gating/check/run_deploy_mnaio.sh]: Ah yes, that should be ``is the deploy action``. I'll have to correct that in master and port it back.

Comment (odyssey4me) [reply gating/check/run_deploy_mnaio.sh]: PR up: https://github.com/rcbops/rpc-openstack/pull/3308

**Pull request merged**